### PR TITLE
PLU-26: [FORMSG-WEBHOOK-5]  backend mutation to register webhook

### PR DESCRIPTION
**Pull request by pregnantboy**
## Implement registerWebhook mutation

Quite straightforward, we want to also check connection is valid before setting.

### CODE REVIEW

File: packages/backend/src/graphql/mutations/register-webhook.ts [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+  }
+  if (!isStillVerified) {
+    throw new Error('Connection is not verified')
+  }
+
+  // Register webhook
+  try {
+    await trigger.registerHook($)
+  } catch (e) {
+    logger.error('Unable to register webhook', e)
```
Comment (ogp-weeloong) [inline packages/backend/src/graphql/mutations/register-webhook.ts:60]: nit liek in #159 - do we want more detailed logging

**Review by ogp-weeloong — APPROVED** [review]
LGTM.

I was a bit worried about introducing a vuln since verify + register is not atomic (e.g. malicious user changes form ID between verify + register by running `updateConnection` mutation).

But I see that FormSG's registerWebhook function also reads from global variable, so we're guaranteed it'll always pick up the verified form ID.

That said, we could accidentally introduce a vuln in the future in other apps, unless we somehow ensure that `registerWebhook` always uses global var instead of something like querying the connection again.

Wondering if a safer method is make `registerWebhook` accept the explicitly verified data as the arguments, instead of making global variable its argument.

**Comment by pregnantboy** [comment]
> LGTM.
> 
> I was a bit worried about introducing a vuln since verify + register is not atomic (e.g. malicious user changes form ID between verify + register by running `updateConnection` mutation).
> 
> But I see that FormSG's registerWebhook function also reads from global variable, so we're guaranteed it'll always pick up the verified form ID.
> 
> That said, we could accidentally introduce a vuln in the future in other apps, unless we somehow ensure that `registerWebhook` always uses global var instead of something like querying the connection again.
> 
> Wondering if a safer method is make `registerWebhook` accept the explicitly verified data as the arguments, instead of making global variable its argument.

Good point. I was initially thinking of calling verify credentials within verify/register webhook to prevent this. But looking at the impact of this, even if an authenticated but malicious user was able to set the webhook for a form that doesnt belong to him, he'd still need the secret key to decrypt the response. Andif he has the secret key, he can just legitimately set the form webhook on plumber as well.

**Comment by pregnantboy** [comment]
> > LGTM.
> > I was a bit worried about introducing a vuln since verify + register is not atomic (e.g. malicious user changes form ID between verify + register by running `updateConnection` mutation).
> > But I see that FormSG's registerWebhook function also reads from global variable, so we're guaranteed it'll always pick up the verified form ID.
> > That said, we could accidentally introduce a vuln in the future in other apps, unless we somehow ensure that `registerWebhook` always uses global var instead of something like querying the connection again.
> > Wondering if a safer method is make `registerWebhook` accept the explicitly verified data as the arguments, instead of making global variable its argument.
> 
> Good point. I was initially thinking of calling verify credentials within verify/register webhook to prevent this. But looking at the impact of this, even if an authenticated but malicious user was able to set the webhook for a form that doesnt belong to him, he'd still need the secret key to decrypt the response. Andif he has the secret key, he can just legitimately set the form webhook on plumber as well.

Of course, this is only in the context of FormSG and might be different for other apps in future.

File: packages/backend/src/graphql/mutations/register-webhook.ts (hunk 1 above)
Comment (pregnantboy) [reply packages/backend/src/graphql/mutations/register-webhook.ts:60]: added in individual (FormSG only for now) registerHook implementation

**Pull request merged**